File: WMSWebAPI/Dal/Producto/clsLnProducto_imagen.cs
using Microsoft.Data.SqlClient;
using System.Data;
using System.Diagnostics;
using WMSWebAPI.Entity.Producto;
public class clsLnProducto_imagen
{
    private static readonly clsInsert Ins = new clsInsert();
    private static readonly clsUpdate Upd = new clsUpdate();

    public static void Cargar(ref clsBeProducto_imagen oBe, DataRow dr)
    {
        oBe.IdProductoImagen = dr.Field<int?>("IdProductoImagen") ?? 0;
        oBe.IdProducto = dr.Field<int?>("IdProducto") ?? 0;
        oBe.Etiqueta = dr.Field<string>("Etiqueta") ?? "";
        oBe.Imagen = dr.Field<byte[]>("Imagen") ?? Array.Empty<byte>();
        oBe.User_agr = dr.Field<string>("user_agr") ?? "";
        oBe.Fec_agr = dr.Field<DateTime?>("fec_agr") ?? DateTime.Now;
    }

    public static bool Existe(int id, SqlConnection conn, SqlTransaction tx)
    {
        const string sql = "SELECT COUNT(1) FROM Producto_imagen WHERE IdProductoImagen = @Id";

        try
        {
            using var cmd = new SqlCommand(sql, conn, tx);
            cmd.Parameters.AddWithValue("@Id", id);
            return Convert.ToInt32(cmd.ExecuteScalar()) > 0;
        }
        catch (Exception ex)
        {
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
    }
    public static int Insertar(IConfiguration config, clsBeProducto_imagen e, SqlConnection? conn = null, SqlTransaction? tx = null)
    {
        bool externa = conn != null && tx != null;
        var lConn = externa ? conn! : new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? lTx = null;
        if (!externa) { lConn.Open(); lTx = lConn.BeginTransaction(IsolationLevel.ReadUncommitted); }

        try
        {
            Ins.Init("producto_imagen");
            Ins.Add("idproductoimagen", "@idproductoimagen", "F");
            Ins.Add("idproducto", "@idproducto", "F");
            Ins.Add("etiqueta", "@etiqueta", "F");
            Ins.Add("imagen", "@imagen", "F");
            Ins.Add("user_agr", "@user_agr", "F");
            Ins.Add("fec_agr", "@fec_agr", "F");

            var sql = Ins.SQL();
            using var cmd = new SqlCommand(sql, lConn, externa ? tx! : lTx!);
            Bind(cmd, e);
            int result = cmd.ExecuteNonQuery();

            if (!externa) lTx?.Commit();
            return result;
        }
        catch (Exception ex)
        {
            if (!externa) lTx?.Rollback();
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
        finally
        {
            if (!externa)
            {
                lTx?.Dispose();
                lConn.Close();
                lConn.Dispose();
            }
        }
    }
    public static int Actualizar(IConfiguration config, clsBeProducto_imagen e, SqlConnection? conn = null, SqlTransaction? tx = null)
    {
        bool externa = conn != null && tx != null;
        var lConn = externa ? conn! : new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? lTx = null;
        if (!externa) { lConn.Open(); lTx = lConn.BeginTransaction(IsolationLevel.ReadUncommitted); }

        try
        {
            Upd.Init("producto_imagen");
            Upd.Add("idproducto", "@idproducto", "F");
            Upd.Add("etiqueta", "@etiqueta", "F");
            Upd.Add("imagen", "@imagen", "F");
            Upd.Add("user_agr", "@user_agr", "F");
            Upd.Add("fec_agr", "@fec_agr", "F");
            Upd.Where("IdProductoImagen = @IdProductoImagen");

            var sql = Upd.SQL();
            using var cmd = new SqlCommand(sql, lConn, externa ? tx! : lTx!);
            Bind(cmd, e);
            int result = cmd.ExecuteNonQuery();

            if (!externa) lTx?.Commit();
            return result;
        }
        catch (Exception ex)
        {
            if (!externa) lTx?.Rollback();
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
        finally
        {
            if (!externa)
            {
                lTx?.Dispose();
                lConn.Close();
                lConn.Dispose();
            }
        }
    }
    public static bool GetSingle(IConfiguration config, ref clsBeProducto_imagen e)
    {
        const string sql = "SELECT * FROM Producto_imagen WHERE IdProductoImagen = @Id";
        using var conn = new SqlConnection(config.GetConnectionString("CST"));
        conn.Open();

        using var tran = conn.BeginTransaction(IsolationLevel.ReadUncommitted);
        try
        {
            using var cmd = new SqlCommand(sql, conn, tran);
            cmd.Parameters.AddWithValue("@Id", e.IdProductoImagen);

            using var adapter = new SqlDataAdapter(cmd);
            var dt = new DataTable();
            adapter.Fill(dt);

            tran.Commit();

            if (dt.Rows.Count == 1)
            {
                Cargar(ref e, dt.Rows[0]);
                return true;
            }

            return false;
        }
        catch (Exception ex)
        {
            tran.Rollback();
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
        finally
        {
            conn.Close();
            conn.Dispose();
        }
    }
    public static List<clsBeProducto_imagen> GetAll(IConfiguration config)
    {
        const string sql = "SELECT * FROM Producto_imagen";
        using var conn = new SqlConnection(config.GetConnectionString("CST"));
        conn.Open();

        using var tran = conn.BeginTransaction(IsolationLevel.ReadUncommitted);
        try
        {
            using var cmd = new SqlCommand(sql, conn, tran);
            using var adapter = new SqlDataAdapter(cmd);
            var dt = new DataTable();
            adapter.Fill(dt);

            tran.Commit();

            var list = new List<clsBeProducto_imagen>();
            foreach (DataRow row in dt.Rows)
            {
                var item = new clsBeProducto_imagen();
                Cargar(ref item, row);
                list.Add(item);
            }

            return list;
        }
        catch (Exception ex)
        {
            tran.Rollback();
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
        finally
        {
            conn.Close();
            conn.Dispose();
        }
    }
    public static DataTable Listar(IConfiguration config)
    {
        const string sql = "SELECT * FROM Producto_imagen";
        using var conn = new SqlConnection(config.GetConnectionString("CST"));
        conn.Open();

        using var tran = conn.BeginTransaction(IsolationLevel.ReadUncommitted);
        try
        {
            using var cmd = new SqlCommand(sql, conn, tran);
            using var adapter = new SqlDataAdapter(cmd);
            var dt = new DataTable();
            adapter.Fill(dt);

            tran.Commit();
            return dt;
        }
        catch (Exception ex)
        {
            tran.Rollback();
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
        finally
        {
            conn.Close();
            conn.Dispose();
        }
    }
    public static int MaxID(IConfiguration config, SqlConnection? conn = null, SqlTransaction? tx = null)
    {
        const string sql = "SELECT ISNULL(MAX(IdProductoImagen), 0) FROM Producto_imagen";
        bool externa = conn != null && tx != null;
        var lConn = externa ? conn! : new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? lTx = null;

        if (!externa) { lConn.Open(); lTx = lConn.BeginTransaction(); }

        try
        {
            using var cmd = new SqlCommand(sql, lConn, externa ? tx! : lTx!);
            var result = cmd.ExecuteScalar();
            if (!externa) lTx?.Commit();
            return Convert.ToInt32(result);
        }
        catch
        {
            if (!externa) lTx?.Rollback();
            throw;
        }
        finally
        {
            if (!externa) lConn.Close();
        }
    }
    private static void Bind(SqlCommand cmd, clsBeProducto_imagen e)
    {
        cmd.Parameters.AddWithValue("@idproductoimagen", e.IdProductoImagen);
        cmd.Parameters.AddWithValue("@idproducto", e.IdProducto);
        cmd.Parameters.AddWithValue("@etiqueta", e.Etiqueta);
        cmd.Parameters.AddWithValue("@imagen", e.Imagen);
        cmd.Parameters.AddWithValue("@user_agr", e.User_agr);
        cmd.Parameters.AddWithValue("@fec_agr", e.Fec_agr);
    }
}

File: TOMIMSV4/DAL/Mantenimientos/Producto/Producto_Imagen/clsLnProducto_imagen.vb
Imports System.Data.SqlClient
Imports System.Reflection

Public Class clsLnProducto_imagen

    Public Shared Sub Cargar(ByRef oBeProducto_imagen As clsBeProducto_imagen, ByRef dr As DataRow)
        Try
            With oBeProducto_imagen
                .IdProductoImagen = IIf(IsDBNull(dr.Item("IdProductoImagen")), 0, dr.Item("IdProductoImagen"))
                .IdProducto = IIf(IsDBNull(dr.Item("IdProducto")), 0, dr.Item("IdProducto"))
                .Etiqueta = IIf(IsDBNull(dr.Item("Etiqueta")), "", dr.Item("Etiqueta"))
                .Imagen = IIf(IsDBNull(dr.Item("Imagen")), Nothing, dr.Item("Imagen"))
                .User_agr = IIf(IsDBNull(dr.Item("user_agr")), "", dr.Item("user_agr"))
                .Fec_agr = IIf(IsDBNull(dr.Item("fec_agr")), Date.Now, dr.Item("fec_agr"))
            End With
        Catch ex As Exception
            Throw New Exception(String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message))
        End Try
    End Sub

    Public Shared Function Insertar(ByRef oBeProducto_imagen As clsBeProducto_imagen, Optional ByVal pConection as SqlConnection = Nothing, Optional Byval pTransaction as SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(connectionString:=Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try

            Ins.Init("producto_imagen")
            Ins.Add("idproductoimagen", "@idproductoimagen", DataType.Parametro)
            Ins.Add("idproducto", "@idproducto", DataType.Parametro)
            Ins.Add("etiqueta", "@etiqueta", DataType.Parametro)
            Ins.Add("imagen", "@imagen", DataType.Parametro)
            Ins.Add("user_agr", "@user_agr", DataType.Parametro)
            Ins.Add("fec_agr", "@fec_agr", DataType.Parametro)

            Dim sp As String = Ins.SQL()
            Dim cmd As New SqlCommand With {.CommandType = CommandType.Text}

            Dim Es_Transaccion_Remota As Boolean = (Not pConection Is Nothing AndAlso Not pTransaction Is Nothing)

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection, pTransaction)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUnCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDPRODUCTOIMAGEN", oBeProducto_imagen.IdProductoImagen))
            cmd.Parameters.Add(New SqlParameter("@IDPRODUCTO", oBeProducto_imagen.IdProducto))
            cmd.Parameters.Add(New SqlParameter("@ETIQUETA", oBeProducto_imagen.Etiqueta))
            cmd.Parameters.Add(New SqlParameter("@IMAGEN", oBeProducto_imagen.Imagen))
            cmd.Parameters.Add(New SqlParameter("@USER_AGR", oBeProducto_imagen.User_agr))
            cmd.Parameters.Add(New SqlParameter("@FEC_AGR", oBeProducto_imagen.Fec_agr))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            cmd.Dispose

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex As Exception
            If Not lTransaction Is Nothing Then lTransaction.Rollback()
            Throw New Exception(String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message))
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close
            If Not lConnection is Nothing Then lConnection.Dispose()
            If Not lTransaction is Nothing Then lTransaction.Dispose()
        End Try

    End Function

    Public Shared Function Actualizar(ByRef oBeProducto_imagen As clsBeProducto_imagen, Optional ByVal pConection as SqlConnection = Nothing, Optional Byval pTransaction as SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(connectionString:=Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try

            Upd.Init("producto_imagen")
            Upd.Add("idproductoimagen", "@idproductoimagen", DataType.Parametro)
            Upd.Add("idproducto", "@idproducto", DataType.Parametro)
            Upd.Add("etiqueta", "@etiqueta", DataType.Parametro)
            Upd.Add("imagen", "@imagen", DataType.Parametro)
            Upd.Add("user_agr", "@user_agr", DataType.Parametro)
            Upd.Add("fec_agr", "@fec_agr", DataType.Parametro)
            Upd.Where("IdProductoImagen = @IdProductoImagen")

            Dim sp As String = Upd.SQL()

            Dim Es_Transaccion_Remota As Boolean = (Not pConection Is Nothing AndAlso Not pTransaction Is Nothing)

            Dim cmd As New SqlCommand With {.CommandType = CommandType.Text}

            If Es_Transaccion_Remota then
                cmd = New SqlCommand(sp, pConection, pTransaction)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUnCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDPRODUCTOIMAGEN", oBeProducto_imagen.IdProductoImagen))
            cmd.Parameters.Add(New SqlParameter("@IDPRODUCTO", oBeProducto_imagen.IdProducto))
            cmd.Parameters.Add(New SqlParameter("@ETIQUETA", oBeProducto_imagen.Etiqueta))
            cmd.Parameters.Add(New SqlParameter("@IMAGEN", oBeProducto_imagen.Imagen))
            cmd.Parameters.Add(New SqlParameter("@USER_AGR", oBeProducto_imagen.User_agr))
            cmd.Parameters.Add(New SqlParameter("@FEC_AGR", oBeProducto_imagen.Fec_agr))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            cmd.Dispose

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex As Exception
            If Not lTransaction Is Nothing Then lTransaction.Rollback()
            Throw New Exception(String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message))
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close
            If Not lConnection is Nothing Then lConnection.Dispose()
            If Not lTransaction is Nothing Then lTransaction.Dispose()
        End Try

    End Function


    Public Shared Function Eliminar(ByRef oBeProducto_imagen As clsBeProducto_imagen, Optional ByVal pConection as SqlConnection = Nothing, Optional Byval pTransaction as SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(connectionString:=Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try

            Const sp As String = " Delete from Producto_imagen" & _
             "  Where(IdProductoImagen = @IdProductoImagen)"

            Dim cmd As New SqlCommand With {.CommandType = CommandType.Text}

            Dim Es_Transaccion_Remota As Boolean = (Not pConection Is Nothing AndAlso Not pTransaction Is Nothing)

            If Es_Transaccion_Remota then
                cmd = New SqlCommand(sp, pConection, pTransaction)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUnCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDPRODUCTOIMAGEN", oBeProducto_imagen.IdProductoImagen))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            cmd.Dispose

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex As Exception
            If Not lTransaction Is Nothing Then lTransaction.Rollback()
            Throw New Exception(String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message))
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close
            If Not lConnection is Nothing Then lConnection.Dispose()
            If Not lTransaction is Nothing Then lTransaction.Dispose()
        End Try

    End Function

    Public Shared Function Listar() As DataTable

        Dim lConnection As New SqlConnection(connectionString:=Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try

            Const sp As String = "SELECT * FROM Producto_imagen"
            lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUnCommitted)
            Dim cmd As New SqlCommand(sp, lConnection, lTransaction) With {.CommandType = CommandType.Text}
            Dim dad As New SqlDataAdapter(cmd)
            Dim dt As New DataTable
            dad.Fill(dt)

            lTransaction.Commit()

            Return dt

        Catch ex As Exception
            If Not lTransaction Is Nothing Then lTransaction.Rollback()
            Throw New Exception(String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message))
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close
            If Not lConnection is Nothing Then lConnection.Dispose()
            If Not lTransaction is Nothing Then lTransaction.Dispose()
        End Try

    End Function

    Public Shared Function Get_All() As List(Of clsBeProducto_imagen)

        Dim lReturnList As New List(Of clsBeProducto_imagen)

        Try

            Const sp As String = "SELECT * FROM Producto_imagen"

            Using lConnection As New SqlConnection(connectionString:=Configuration.ConfigurationManager.AppSettings("CST"))

                lConnection.Open()

                Using lTransaction As SqlTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUnCommitted)

                    Using lDTA As New SqlDataAdapter(sp, lConnection)

                        lDTA.SelectCommand.CommandType = CommandType.Text
                        lDTA.SelectCommand.Transaction = lTransaction
                        Dim lDataTable As New DataTable
                        lDTA.Fill(lDataTable)

                        Dim vBeProducto_imagen As New clsBeProducto_imagen

                        For Each dr As DataRow In lDataTable.Rows
                            vBeProducto_imagen = New clsBeProducto_imagen()
                            Cargar(vBeProducto_imagen, dr)
                            lReturnList.Add(vBeProducto_imagen)
                        Next

                    End Using

                    lTransaction.Commit()

                End Using

                lConnection.Close()

            End Using

            Return lReturnList

        Catch ex As Exception
            Throw New Exception(String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message))
        End Try

    End Function

    Public Shared Sub GetSingle(ByRef pBeProducto_imagen As clsBeProducto_imagen)

        Try

            Const sp As String = "SELECT * FROM Producto_imagen" & _
            " Where(IdProductoImagen = @IdProductoImagen)"


            Using lConnection As New SqlConnection(connectionString:=Configuration.ConfigurationManager.AppSettings("CST"))

                lConnection.Open()

                Using lTransaction As SqlTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUnCommitted)

                    Using lDTA As New SqlDataAdapter(sp, lConnection)

                        lDTA.SelectCommand.CommandType = CommandType.Text
                        lDTA.SelectCommand.Transaction = lTransaction
                        Dim lDataTable As New DataTable
                        lDTA.Fill(lDataTable)

                        Dim vBeProducto_imagen As New clsBeProducto_imagen

                        If lDataTable IsNot Nothing AndAlso lDataTable.Rows.Count > 0 Then
                            Cargar(vBeProducto_imagen, lDataTable.Rows(0))
                        End If

                    End Using

                    lTransaction.Commit()

                End Using

                lConnection.Close()

            End Using

        Catch ex As Exception
            Throw New Exception(String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message))
        End Try

    End Sub

    Public Shared Function MaxID() as Integer

        Try

            Dim lMax As Integer = 0

            Const sp As String = "SELECT ISNULL(Max(IdProductoImagen),0) FROM Producto_imagen"

            Using lConnection As New SqlConnection(connectionString:=Configuration.ConfigurationManager.AppSettings("CST"))

                lConnection.Open()

                Using lTransaction As SqlTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUnCommitted)

                    Using lCommand As New SqlCommand(sp, lConnection, lTransaction) With {.CommandType = CommandType.Text}

                        Dim lReturnValue As Object = lCommand.ExecuteScalar()
                        If lReturnValue IsNot DBNull.Value AndAlso lReturnValue IsNot Nothing Then
                            lMax = CInt(lReturnValue)
                        End If

                    End Using

                    lTransaction.Commit()

                End Using

                lConnection.Close()

            End Using

            Return lMax

        Catch ex As Exception
            Throw New Exception(String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message))
        End Try

    End Function

End Class
